### ISSUE TITLE
Adding modules to app engine server causes hang

### ISSUE THREAD

**Issue by branflake2267** [GoogleCloudPlatform/google-cloud-eclipse#914]
Adding multiple module to app engine causes hang. No evident logging can be seen. No console output that I can provide to help in rcp debuggin. 

hangs on this dialog
![screen shot 2016-10-29 at 10 05 55 pm](https://cloud.githubusercontent.com/assets/1326504/19834625/396627a2-9e24-11e6-899c-13b6c083aaca.png)

Replicate
1. create gwt project from tbroyer multi module web app  (https://github.com/tbroyer/gwt-maven-archetypes - directions) (`modular-webapp` archetype selected)
2. change pom.xml java source/target to 1.7. 
3. select app engine facet for modules

Modules, all have app engine facet selected
![screen shot 2016-10-29 at 10 12 42 pm](https://cloud.githubusercontent.com/assets/1326504/19834640/d2662ea2-9e24-11e6-99d5-111df4dac52c.png)

versions:
![screen shot 2016-10-29 at 10 11 31 pm](https://cloud.githubusercontent.com/assets/1326504/19834639/ada4ed74-9e24-11e6-97d7-76c74cd0ba6b.png)

**Comment by briandealwis** [comment]
Seems to be getting into an infinite loop called where `ServerWorkingCopy.resetPublishState()` is calling into itself:
```
Thread 85883: (state = BLOCKED)
 - org.eclipse.osgi.internal.framework.FilterImpl$Parser.parse_filtercomp() @bci=94, line=1469 (Compiled frame)
 - org.eclipse.osgi.internal.framework.FilterImpl$Parser.parse_filter() @bci=58, line=1435 (Compiled frame)
 - org.eclipse.osgi.internal.framework.FilterImpl$Parser.parse() @bci=1, line=1414 (Compiled frame)
 - org.eclipse.osgi.internal.framework.FilterImpl.newInstance(java.lang.String, boolean) @bci=9, line=147 (Compiled frame)
 - org.eclipse.osgi.internal.framework.FilterImpl.newInstance(java.lang.String) @bci=2, line=143 (Compiled frame)
 - org.eclipse.osgi.internal.container.Capabilities$NamespaceSet.findCapabilities(org.osgi.resource.Requirement) @bci=81, line=120 (Compiled frame)
 - org.eclipse.osgi.internal.container.Capabilities.findCapabilities(org.osgi.resource.Requirement) @bci=29, line=264 (Compiled frame)
 - org.eclipse.osgi.container.ModuleDatabase.findCapabilities(org.osgi.resource.Requirement) @bci=9, line=826 (Compiled frame)
 - org.eclipse.osgi.container.ModuleContainer$ContainerWiring.findProviders(org.osgi.resource.Requirement) @bci=8, line=1312 (Compiled frame)
 - org.eclipse.osgi.internal.framework.legacy.PackageAdminImpl.getBundles(java.lang.String, java.lang.String) @bci=163, line=181 (Compiled frame)
 - org.eclipse.core.internal.runtime.InternalPlatform.getBundle(java.lang.String) @bci=14, line=182 (Compiled frame)
 - org.eclipse.core.runtime.Platform.getBundle(java.lang.String) @bci=4, line=1398 (Compiled frame)
 - org.eclipse.core.expressions.PropertyTester.isDeclaringPluginActive() @bci=14, line=119 (Compiled frame)
 - org.eclipse.core.internal.expressions.Property.isDeclaringPluginActive() @bci=4, line=45 (Compiled frame)
 - org.eclipse.core.internal.expressions.Property.isValidCacheEntry(boolean) @bci=30, line=52 (Compiled frame)
 - org.eclipse.core.internal.expressions.TypeExtensionManager.getProperty(java.lang.Object, java.lang.String, java.lang.String, boolean) @bci=67, line=104 (Compiled frame)
 - org.eclipse.core.internal.expressions.TestExpression.evaluate(org.eclipse.core.expressions.IEvaluationContext) @bci=79, line=97 (Compiled frame)
 - org.eclipse.core.internal.expressions.CompositeExpression.evaluateOr(org.eclipse.core.expressions.IEvaluationContext) @bci=43, line=68 (Compiled frame)
 - org.eclipse.core.internal.expressions.OrExpression.evaluate(org.eclipse.core.expressions.IEvaluationContext) @bci=2, line=22 (Compiled frame)
 - org.eclipse.core.internal.expressions.CompositeExpression.evaluateAnd(org.eclipse.core.expressions.IEvaluationContext) @bci=43, line=53 (Compiled frame)
 - org.eclipse.core.internal.expressions.EnablementExpression.evaluate(org.eclipse.core.expressions.IEvaluationContext) @bci=14, line=55 (Compiled frame)
 - org.eclipse.wst.server.core.internal.ModuleFactory.evalEnablementExpression(org.eclipse.core.expressions.IEvaluationContext, org.eclipse.core.expressions.Expression) @bci=6, line=186 (Compiled frame)
 - org.eclipse.wst.server.core.internal.ModuleFactory.isEnabled(org.eclipse.core.resources.IProject, org.eclipse.core.runtime.IProgressMonitor) @bci=25, line=194 (Compiled frame)
 - org.eclipse.wst.server.core.ServerUtil.getModules(org.eclipse.core.resources.IProject) @bci=50, line=97 (Compiled frame)
 - org.eclipse.wst.web.internal.deployables.FlatComponentDeployable.gatherModuleReference(org.eclipse.wst.common.componentcore.resources.IVirtualComponent, org.eclipse.wst.common.componentcore.internal.flat.IChildModuleReference) @bci=42, line=328 (Compiled frame)
 - org.eclipse.jst.j2ee.internal.deployables.J2EEFlexProjDeployable.gatherModuleReference(org.eclipse.wst.common.componentcore.resources.IVirtualComponent, org.eclipse.wst.common.componentcore.internal.flat.IChildModuleReference) @bci=12, line=126 (Compiled frame)
 - org.eclipse.jst.jee.internal.deployables.JEEFlexProjDeployable.gatherModuleReference(org.eclipse.wst.common.componentcore.resources.IVirtualComponent, org.eclipse.wst.common.componentcore.internal.flat.IChildModuleReference) @bci=12, line=56 (Compiled frame)
 - org.eclipse.wst.web.internal.deployables.FlatComponentDeployable.getModules() @bci=43, line=265 (Compiled frame)
 - com.google.cloud.tools.eclipse.appengine.localserver.server.LocalAppEngineServerDelegate.getChildModules(org.eclipse.wst.server.core.IModule[]) @bci=76, line=86 (Compiled frame)
 - org.eclipse.wst.server.core.internal.Server.getChildModules(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=88, line=2637 (Compiled frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetPublishState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=9, line=647 (Compiled frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetPublishState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=59, line=654 (Compiled frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetPublishState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=59, line=654 (Compiled frame)
```
_hundreds of lines elided_
```
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetPublishState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=59, line=654 (Interpreted frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetPublishState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=59, line=654 (Interpreted frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.resetState(org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=77, line=628 (Interpreted frame)
 - org.eclipse.wst.server.core.internal.ServerWorkingCopy.modifyModules(org.eclipse.wst.server.core.IModule[], org.eclipse.wst.server.core.IModule[], org.eclipse.core.runtime.IProgressMonitor) @bci=138, line=563 (Interpreted frame)
 - org.eclipse.wst.server.ui.internal.wizard.WizardTaskUtil.modifyModules(java.util.List, java.util.List, org.eclipse.wst.server.core.TaskModel, org.eclipse.core.runtime.IProgressMonitor) @bci=179, line=226 (Interpreted frame)
 - org.eclipse.wst.server.ui.internal.wizard.fragment.ModifyModulesWizardFragment.performFinish(org.eclipse.core.runtime.IProgressMonitor) @bci=26, line=133 (Interpreted frame)
 - org.eclipse.wst.server.ui.internal.wizard.TaskWizard.executeTask(org.eclipse.wst.server.ui.wizard.WizardFragment, byte, org.eclipse.core.runtime.IProgressMonitor) @bci=12, line=268 (Interpreted frame)
 - org.eclipse.wst.server.ui.internal.wizard.TaskWizard$2.run(org.eclipse.core.runtime.IProgressMonitor) @bci=151, line=215 (Interpreted frame)
 - org.eclipse.core.internal.resources.Workspace.run(org.eclipse.core.resources.IWorkspaceRunnable, org.eclipse.core.runtime.jobs.ISchedulingRule, int, org.eclipse.core.runtime.IProgressMonitor) @bci=80, line=2241 (Interpreted frame)
 - org.eclipse.core.internal.resources.Workspace.run(org.eclipse.core.resources.IWorkspaceRunnable, org.eclipse.core.runtime.IProgressMonitor) @bci=8, line=2225 (Interpreted frame)
 - org.eclipse.wst.server.ui.internal.wizard.page.WorkspaceRunnableAdapter.run(org.eclipse.core.runtime.IProgressMonitor) @bci=8, line=35 (Interpreted frame)
 - org.eclipse.jface.operation.ModalContext$ModalContextThread.run() @bci=15, line=119 (Interpreted frame)

**Comment by briandealwis** [comment]
This is an interest test case for a couple of reasons.  This test has the `-server` project include the `-shared` and `-client` projects (both are utility modules).

First, it tickles an unexpected pathological case.  Other WTP server delegates implementations of `#getChildModules(IModule[] modules)` only process the provided _modules_ if _modules.length == 1_, whereas our `LocalAppEngineServerDelegate` does not.

There is an additional problems that the classes from `-shared` are being published into the module root rather than in `WEB-INF/classes`, nor as a jar in `WEB-INF/lib`.  `-shared` is a utility module. I think this is #136.

**Comment by briandealwis** [comment]
`ServerDelegate#getChildModules(IModule[])` is passed an array of `IModule`s which represent the current path through the module hierarchy.  The last module in the array is the real module under consideration.  This isn't well-documented, but it can be seen from how `ServerWorkingCopy#resetState()` and `#resetPublisState()` build up the `IModule` arrays passed in.

**Comment by elharo** [comment]
This is bad. If we can get a new release cut that fixes this before we officially launch next week, I'd like to do that.

**Comment by patflynn** [comment]
Question: Did GPE support multi-module local development?

**Comment by branflake2267** [comment]
Not about GPE, but I did test this with Tomcat and this config works with it. Not sure that has bearing on this. Although I'm advertising this config at GWT Con as it's one of the new preferred ways of building a GWT App. That said, the archetype sets up Tomcat, I'd like to provide an App engine version at some point when the issue is solved.

**Comment by briandealwis** [comment]
> @patflynn: Question: Did GPE support multi-module local development?

From the [original bug](https://code.google.com/p/googleappengine/issues/detail?id=10652&q=wtp&colspec=ID%20Type%20Component%20Status%20Stars%20Summary%20Language%20Priority%20Owner%20Log) linked to in #136, I'm guessing "no" but it was desired.  Just looking into #136 now.

**Comment by branflake2267** [comment]
Could you update the gcloud alpha update repo so I could test this?

**Comment by branflake2267** [comment]
FYI... I pulled the latest commits, built and installed. I was able to add the multi modules, but there were other issues that popped up.

**Comment by briandealwis** [comment]
I think you're hitting #136 with the `-server` requiring `-shared` and `-client`.

**Comment by briandealwis** [comment]
Sadly even with the fix from #984, that `modular-webapp` archetype doesn't run unmodified as it requires Java 1.8.
